Core/Src/main.c

```diff
@@ -0,0 +1,595 @@
+
+  /*Configure GPIO pins : PC2 PC3 */
+  GPIO_InitStruct.Pin = GPIO_PIN_2|GPIO_PIN_3;
+  GPIO_InitStruct.Mode = GPIO_MODE_OUTPUT_PP;
+  GPIO_InitStruct.Pull = GPIO_NOPULL;
+  GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_LOW;
+  HAL_GPIO_Init(GPIOC, &GPIO_InitStruct);
+
+  /*Configure GPIO pins : LD2_Pin PA8 */
+  GPIO_InitStruct.Pin = LD2_Pin|GPIO_PIN_8;
+  GPIO_InitStruct.Mode = GPIO_MODE_OUTPUT_PP;
+  GPIO_InitStruct.Pull = GPIO_NOPULL;
+  GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_LOW;
+  HAL_GPIO_Init(GPIOA, &GPIO_InitStruct);
+
+  /*Configure GPIO pins : PA9 PA11 PA12 */
+  GPIO_InitStruct.Pin = GPIO_PIN_9|GPIO_PIN_11|GPIO_PIN_12;
+  GPIO_InitStruct.Mode = GPIO_MODE_IT_RISING;
+  GPIO_InitStruct.Pull = GPIO_NOPULL;
+  HAL_GPIO_Init(GPIOA, &GPIO_InitStruct);
+
+  /* EXTI interrupt init*/
+  HAL_NVIC_SetPriority(EXTI9_5_IRQn, 0, 0);
+  HAL_NVIC_EnableIRQ(EXTI9_5_IRQn);
+
+  HAL_NVIC_SetPriority(EXTI15_10_IRQn, 0, 0);
+  HAL_NVIC_EnableIRQ(EXTI15_10_IRQn);
+
+/* USER CODE BEGIN MX_GPIO_Init_2 */
+/* USER CODE END MX_GPIO_Init_2 */
+}
+
+/* USER CODE BEGIN 4 */
+//Motorfeedback should read the yellow and white wires, and determine the motor speed,
+//incremental position and direction from the 2 square waves.
+/*
+void motorfeedback(){
+
+	int Yellowsig = HAL_GPIO_ReadPin(GPIOA, GPIO_PIN_12);//digitalRead(encoder0pinA);
+	 if((YellowsigLast == 0) && Yellowsig == 1)
+	 {
+	   int Whitesig = HAL_GPIO_ReadPin(GPIOA, GPIO_PIN_12); //digitalRead(encoder0pinB);
+	   if(Whitesig == 0 && Direction)
+	   {
+	     Direction = 0; //Reverse/false
+	   }
+	   else if(Whitesig == 1 && Direction == 0)
+	   {
+	     Direction = 1;  //Forward/true
+	   }
+	 }
+	 YellowsigLast = Yellowsig;
+
+	 if(Direction == 0)  duration++;
+	 else  duration--;
+	 printf("%d", duration); //see feedback via console
+}
+*/
+
+int IntegralControl(int errorterm) {
+
+	for (j = 29; j > 0; j--) { //moving down the array terms, with the oldest being 10 and youngest 0
+		IntArray[j]=IntArray[j-1];
+	}
+	IntArray[0] = errorterm;
+
+	sum = 0;
+	for(k = 0; k < 29; k++) { //summing the entire array
+		sum = sum + IntArray[k]; // same as sum += arr[i];
+	}
+
+	IntegralTerm = IntTerm*sum;
+
+	return IntegralTerm;
+}
+
+int DerivativeControl(int errorterm) {
+
+	for (f = 29; f > 0; f--) { //moving down the array terms, with the oldest being 29 and youngest 0
+		DerArray[f]=DerArray[f-1];
+	}
+	DerArray[0] = errorterm;
+
+	//rough approximation of slope by dividing average of first 15 errorterms (numerator)
+	//by the average of the last 15 errorterms (denominator).
+	numerator = 0;
+	denominator = 0;
+	for(z = 0; z < 29; z++) {
+		if(z < 15) {
+			numerator = numerator + DerArray[z];
+		}
+		numerator = numerator/5;
+
+		if(z >= 15) {
+			denominator = denominator + DerArray[z];
+		}
+		denominator = denominator/5;
+
+		if (denominator == 0) {
+			denominator = 1; //preventing possibility of divide by zero error
+		}
+
+		slope = numerator/denominator;
+
+	}
```

Comment:
See my comment for the integral calculation. You can probably do something similar here to reduce the number of operations you have to perform.